src/HDF5.jl

```diff
@@ -277,18 +281,27 @@ const HDF5Scalar = Union{HDF5BitsKind, HDF5ReferenceObj}
 const ScalarOrString = Union{HDF5Scalar, String}
 
 # It's not safe to use particular id codes because these can change, so we use characteristics of the type.
-const hdf5_type_map = Dict(
-    (H5T_INTEGER, H5T_SGN_2, Csize_t(1)) => Int8,
-    (H5T_INTEGER, H5T_SGN_2, Csize_t(2)) => Int16,
-    (H5T_INTEGER, H5T_SGN_2, Csize_t(4)) => Int32,
-    (H5T_INTEGER, H5T_SGN_2, Csize_t(8)) => Int64,
-    (H5T_INTEGER, H5T_SGN_NONE, Csize_t(1)) => UInt8,
-    (H5T_INTEGER, H5T_SGN_NONE, Csize_t(2)) => UInt16,
-    (H5T_INTEGER, H5T_SGN_NONE, Csize_t(4)) => UInt32,
-    (H5T_INTEGER, H5T_SGN_NONE, Csize_t(8)) => UInt64,
-    (H5T_FLOAT, nothing, Csize_t(4)) => Float32,
-    (H5T_FLOAT, nothing, Csize_t(8)) => Float64,
-)
+function _hdf5_type_map(class_id, is_signed, native_size)
+    if class_id == H5T_INTEGER
+        if is_signed == H5T_SGN_2
+            native_size === Csize_t(1) ? (return Int8) :
+            native_size === Csize_t(2) ? (return Int16) :
+            native_size === Csize_t(4) ? (return Int32) :
+            native_size === Csize_t(8) ? (return Int64) :
+            throw(KeyError(class_id, is_signed, native_size))
```

Comment:
I'm not sure this line (and similar in the other branches) is actually necessary — the final check size-8 case in principle could just be the else-case of the size-4 ternary. Thoughts?